trace.go

```diff
@@ -0,0 +1,336 @@
+package pubsub
+
+import (
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	pb "github.com/libp2p/go-libp2p-pubsub/pb"
+)
+
+// Generic event tracer interface
+type EventTracer interface {
+	Trace(evt interface{})
```

Comment:
`pb.Message`?

Reply:
Yeah, we could do that to avoid the cast.

Reply:
done, it takes a `*pb.TraceEvent` now.